docs/interfaces.md

```diff
@@ -0,0 +1,568 @@
+```
+
+# ECC
+
+## Interface:
+The ECC interface is implemented through a "normal" chaincode interface. Methods with leading underscores
+are treated as FPC commands. Normal `invoke` invocations are forwarded to a FPC chaincode enclave.
+
+```go
+// chaincode interface (exposed to admin/clients) implemented by invoke
+func Invoke(stub shim.ChaincodeStubInterface) pb.Response {}
+
+// functions below are dispatched by Invoke implementation
+
+
+// triggered by an admin
+func initEnclave(params CC_Parameters) (Credentials, error) {}
+
+// key generatation 
+func generateCCKeys(params CC_Parameters) (SignedCCKeyRegistrationMessage, error) {}
+
+// key distribution
+func exportCCKeys(credentals Credentials) (SignedExportMessage, error) {}
+func importCCKeys(params CC_Parameters) (SignedCCKeyRegistrationMessage, error) {}
+
+// chaincode invoke
+func chaincodeInvoke(transactionProposal []byte, ctx interface{}) (response, error) {}
+```
+
+```go 
+// Storage interface
+// TODO how to store sealedState, where and under what identifier (e.g. chaincode_id);
+load(chaincode_id string) (sealedState []byte, err error)
+store(chaincode_id string, sealedState []byte) error
+```
+
+This interface is implemented by ECC to let a chaincode enclave call into the peer
+
+```c++
+// Provided/implemented by a common logging module
+// interface exposed to TLCC_enclave via cgo
+void ocall_print_string([in, string] const char *str);
+```
+
+```c++
+// state access
+// TBD does this key already contain the chaincode_id/namespace?
+void ocall_get_state(
+        [in, string] const char *key, 
+        [out, size=max_val_len] uint8_t *val, uint32_t max_val_len, [out] uint32_t *val_len,
+        [in, out] sgx_cmac_128bit_tag_t *cmac,
+        [user_check] void *u_shim_ctx);
+
+void ocall_put_state(
+        [in, string] const char *key,
+        [in, size=val_len] uint8_t *val, uint32_t val_len,
+        [user_check] void *u_shim_ctx);
+
+void ocall_get_state_by_partial_composite_key(
+        [in, string] const char *comp_key,
+        [out, size=max_len] uint8_t *values, uint32_t max_len, [out] uint32_t *values_len,
+        [in, out] sgx_cmac_128bit_tag_t *cmac,
+        [user_check] void *u_shim_ctx);
+
+```
+
+```c++
+// verify msp with the help of tlcc
+void ocall_tlcc_request();
+```
+
+More details on the storage module in [#421](https://github.com/hyperledger-labs/fabric-private-chaincode/issues/421)
+```c++
+// Provided/implemented by a common sealed storage module
+void ocall_load(
+        [in, string] const char *str, 
+        [out, size=max_data_len] uint8_t *data, uint32_t max_data_len
+        [out] uint32_t *data_len);
+
+void ocall_store(
+        [in, string] const char *str, 
+        [in, size=data_len] uint8_t *data, uint32_t data_len);
+```
+
+
+Additionally, ECC includes a common attestation module that abstracts
+
+```c++
+// attestation
+void ocall_get_spid(); // TBD: is spid only needed for EPID?
```

Comment:
I have some doubts on the implementation of this ocall.
In a real deployment, there will likely be one spid per mspid. With such an ocall, the only way (that comes to mind) to retrieve the spid is to leverage the external builder to grab a local spid packaged with the chaincode. So this imposes to package the chaincode in a certain way. Also, if ever the package has to be shared, it might have to be repackaged with a new spid.

For this reason I had originally proposed to send the spid in an Aux_Data field (i.e., a json blob) at create time in the invoke. The advantages are: the caller (i.e., the admin) knows the spid associated to its mspid; the json blob can be repurposed for different attestation types, and in that case it will simply not include the spid; finally, the spid does not ship with the package, hence it is separate from the chaincode and even the peer.

Reply:
Additionally, the json blob can include the sig_rl, which are currently hard-coded.

Reply:
I removed the details of the attestation module used by ECC

Reply:
i would also remove `ocall_create_evidence`. Whether such a thing exists as ecall or just as a library function should really be defined by #412?

Reply:
`ocall_create_evidence` removed